PP-971: MIN_STACK_LIMIT for stacksize not working

bremanandjk opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID : PP-971
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-971](https://pbspro.atlassian.net/browse/PP-971)**

#### Problem description
* The minimum ulimit for stacksize(MIN_STACK_LIMIT) is not working for PBS Pro version 14.x.x and sets this is “unlimited” for pbs_mom component.

#### Cause / Analysis
* Introduction of "<Python.h>" in mom_main.c makes pbs_mom to build in 64 bit mode due to enabling 64 bit macro "RLIM64_INFINITY".

* This third party header "python.h" has "<limits.h>", which defines the "__USE_LARGEFILE64", this again defines the "RLIM64_INFINITY" macro at usr/include/bits/resource.h.

#### Solution description
* The solution would be to set the stack limits to "MIN_STACK_LIMIT" value in 64 bit code path too.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [x] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

bremanandjk commented:
Thanks @sid2364 for your approvals. 
@PBSPro/pbspro-maintainers please take this PR further, Thanks

arungrover commented on src/resmom/mom_main.c:
```diff
@@ -8309,15 +8309,39 @@ main(int argc, char *argv[])
-		(void)getrlimit64(RLIMIT_STACK, &orig_stack_size); /* get for later */
-		(void)setrlimit64(RLIMIT_STACK, &rlimit);
+		if (getrlimit64(RLIMIT_STACK, &orig_stack_size) != -1) {
+			if((orig_stack_size.rlim_cur != RLIM_INFINITY) && (orig_stack_size.rlim_cur < MIN_STACK_LIMIT)) {
```
will it be RLIM_INFINITY or RLIM64_INFINITY for 64 bit machines?

bremanandjk replied:
Thanks for your reviews @arungrover 
Good catch, I have addressed your comment.

arungrover commented on src/resmom/mom_main.c:
```diff
@@ -8309,15 +8309,39 @@ main(int argc, char *argv[])
+		if (getrlimit64(RLIMIT_STACK, &orig_stack_size) != -1) {
+			if((orig_stack_size.rlim_cur != RLIM_INFINITY) && (orig_stack_size.rlim_cur < MIN_STACK_LIMIT)) {
+				rlimit.rlim_cur = MIN_STACK_LIMIT;
+				rlimit.rlim_max = MIN_STACK_LIMIT;
```
why do we set rlim_max to MIN_STACK_LIMIT as well?

bremanandjk replied:
rlim_max is a hard limit (ceiling for rlim_cur) & This hard limit acts as a ceiling for the soft limit. 
 A privileged process may make arbitrary changes to either limit value. 

Reason is just to make the stack size fixed to "MIN_STACK_LIMIT" and not allowing to use beyond this value.
 
Info taken from GNU/Linux man page of "setrlimit", their sample code insist to do so. [man page](http://man7.org/linux/man-pages/man2/getrlimit.2.html)

bremanandjk commented:
I have build and tested the stack size of all PBS Pro daemons after addressing review comments.

[root@science pbspro]# cat /proc/$(pgrep pbs_mom)/limits |grep stack
Max stack size            16777216             16777216             bytes     
[root@science pbspro]# cat /proc/$(pgrep pbs_server)/limits |grep stack
Max stack size            16777216             16777216             bytes     
[root@science pbspro]# cat /proc/$(pgrep pbs_sched)/limits |grep stack
Max stack size            16777216             16777216             bytes     
[root@science pbspro]# cat /proc/$(pgrep pbs_comm)/limits |grep stack
Max stack size            16777216             16777216             bytes     
[root@science pbspro]#

arungrover commented on src/resmom/mom_main.c:
```diff
@@ -8309,15 +8309,39 @@ main(int argc, char *argv[])
-		(void)getrlimit64(RLIMIT_STACK, &orig_stack_size); /* get for later */
-		(void)setrlimit64(RLIMIT_STACK, &rlimit);
+		if (getrlimit64(RLIMIT_STACK, &orig_stack_size) != -1) {
+			if((orig_stack_size.rlim_cur != RLIM64_INFINITY) && (orig_stack_size.rlim_cur < MIN_STACK_LIMIT)) {
```
Can it happen that rlim_cur is between RLIM64_INFINITY and MIN_STACK_LIMIT? do we not want to set rlim_max in that case? Also, if our intent is to set stack limit to MIN_STACK_LIMIT then why don't we set it even when rlim_cur is RLIM64_INFINITY?
I'm still confused about why we are enforcing the rlim_max and rlim_cur both. Should we not be just worried about setting rlim_max to something and leave rlim_cur?

bremanandjk replied:
Can it happen that rlim_cur is between RLIM64_INFINITY and MIN_STACK_LIMIT?

>*Brem*>> A value of RLIM64_INFINITY indicating no limit.(i:e FULL value; equivalent to 0xffffffffffffffffuLL). Only when process running under modified session by setting limit(ulimit -s unlimited) will be rlim_cur = "RLIM64_INFINITY". 
>By default, most GNU/Linux distribution edicts  8MB. With this conditional expression, we are confirming the **there is limit set for stack size from user system && its value less then "Minimum guarantee stack limit for pbs_mom"**. Only when this condition passes, we wanted to boost the stack limit to expected value 16MB(MIN_STACK_LIMIT). 
As we are setting this stack size value before pbs process become daemon, thus next "fork & exec" will enforce this expected stack size to actual pbs_daemon.

And, this is the condition used across all our PBS Pro daemons. 
 
I'm still confused about why we are enforcing the rlim_max and rlim_cur both. Should we not be just worried about setting rlim_max to something and leave rlim_cur?

>*Brem*>> Its must to set rlim_cur & rlim_max, setrlimit syscall fails if we ignore rlim_cur parameter.

arungrover replied:
Got it, makes to set the rlimit to 16MB when it is less than that. 
Another question - Is it okay to leave stack limit set to RLIM64_INFINITY? (if it is set by default)?
Will having a value of INFINITY affects the child processes or threads that this process spawns?

bremanandjk replied:
*Good question,*  👍 
By default, right now none of the systems provides INFINITY stack size to protect the OS for following reason.
>Programs that have a legitimate reason to need more stack are rare. if infinite loop happens to contain a recursive function call, the stack would quickly eat all the available memory. 

As you have mentioned, process/threads it create inherits the same stack limit from parent, By saying that, current logic will help admin to set the desired value(>16MB), if any of the JOB needs it. Restricting to the strict size will make the application (PBS Job) will shoot it by itself due to stack crunch.

The corner case of INFINITY stack size happen only if the system admin sets it by __intentional__. Anyway, here admin know what he is doing and expects to set unlimited for all applications including PBS Pro. I feel it __okay__ to adhere the system settings just like other process. My understanding is, we should avoid the case where PBS Pro crashs due to minimum stack size(boosted to 16mb now) and should provide a way for JOB to grow to the _desired_ value.(if admin set more than 16mb, we wont do any change in it). 

I have shared my thoughts, please throw your valuable suggestions on this, I am open to accept new changes. Thanks!

arungrover replied:
Well, I thought having a large stack size for a process affects the stack size of the threads it spawns. In any case, I overlooked the fact that admin can set ulimit to something between 16MB to INFINITY (if they really want to). 
I sign-off on the change. Please squash and rebase

bremanandjk commented:
Sorry for late reply, 

Thanks @arungrover for your approval & deep reviews, I have squashed my commits to one. 
@sid2364 I have made a change after your approval, please have a look again, Thanks

arungrover commented:
@bremanandjk it is actually my mistake that I told you to squash. But, if you need another reviewer to sign-off it is better we don't squash because it becomes hard for them to review the delta.
Note to myself - don't forget about the second reviewer :)

bremanandjk commented:
@arungrover , Thanks for your note, 
this time, I have mentioned the delta change to second reviewer and got approval, From next time, me too will keep in my mind about second reviewer.

arungrover commented:
Merged
